internal/sops/azkv/keysource.go

```diff
@@ -0,0 +1,195 @@
+)
+
+// MasterKey is an Azure Key Vault key used to encrypt and decrypt SOPS' data key.
+// The underlying authentication token can be configured using AADSettings.
+type MasterKey struct {
+	VaultURL string
+	Name     string
+	Version  string
+
+	EncryptedKey string
+
+	token *adal.ServicePrincipalToken
+}
+
+// LoadAADSettingsFromBytes attempts to load the given bytes into the given AADSettings.
+// By first decoding it if UTF-16, and then unmarshalling it into the given struct.
+// It returns an error for any failure.
+func LoadAADSettingsFromBytes(b []byte, s *AADSettings) error {
+	b, err := decode(b)
+	if err != nil {
+		return fmt.Errorf("failed to decode Azure authentication file bytes: %w", err)
+	}
+	if err = json.Unmarshal(b, s); err != nil {
+		err = fmt.Errorf("failed to unmarshal Azure authentication file: %w", err)
+	}
+	return err
+}
+
+// AADSettings contains the selection of fields from an Azure authentication file
+// required for Active Directory authentication.
+//
+// It is based on the unpublished contract in
+// https://github.com/Azure/go-autorest/blob/c7f947c0610de1bc279f76e6d453353f95cd1bfa/autorest/azure/auth/auth.go#L331-L342,
+// which seems to be due to an assumption of configuration through environment
+// variables over file based configuration.
+type AADSettings struct {
+	ClientID                string `json:"clientId"`
+	ClientSecret            string `json:"clientSecret"`
+	TenantID                string `json:"tenantId"`
+	ActiveDirectoryEndpoint string `json:"activeDirectoryEndpointUrl"`
+}
+
+// SetToken configures the token on the given MasterKey using the AADSettings.
+func (s *AADSettings) SetToken(key *MasterKey) error {
+	if s == nil {
+		return nil
+	}
+	config, err := adal.NewOAuthConfig(s.GetAADEndpoint(), s.TenantID)
+	if err != nil {
+		return err
+	}
+	if key.token, err = adal.NewServicePrincipalToken(*config, s.ClientID, s.ClientSecret,
+		azure.PublicCloud.ResourceIdentifiers.KeyVault); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetAADEndpoint returns the ActiveDirectoryEndpoint, or the Azure Public Cloud
+// default.
+func (s *AADSettings) GetAADEndpoint() string {
+	if s.ActiveDirectoryEndpoint != "" {
+		return s.ActiveDirectoryEndpoint
+	}
+	return azure.PublicCloud.ActiveDirectoryEndpoint
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds.
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key.
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Encrypt takes a SOPS data key, encrypts it with Key Vault and stores the result in the EncryptedKey field.
+func (key *MasterKey) Encrypt(plaintext []byte) error {
+	c := newKeyvaultClient(key.authorizer())
+
+	data := base64.RawURLEncoding.EncodeToString(plaintext)
+	p := keyvault.KeyOperationsParameters{Value: &data, Algorithm: keyvault.RSAOAEP256}
```

Comment:
Might not be relevant today but hard coding the key algo may hurt us in the future.

Reply:
The keysource (like the other keysources), are forks of the SOPS native ones but with modifications to ensure isolation. This setting is thus aligned with SOPS (non-configurable) default behavior.